### PR TITLE
EKF3: Fix covariance matrix stability issues

### PR DESCRIPTION
**Pull request by priseborough**
This PR is a continuation of https://github.com/ArduPilot/ardupilot/pull/6466 and fixes the issues logged by https://github.com/ArduPilot/ardupilot/issues/16791 and https://github.com/ArduPilot/ardupilot/issues/16094.

It fixes an issue created when the vehicle is on ground not moving and the combination of aiding sensor update rate and reported observation accuracy can cause state variances to reduce to a level where the covariance matrix processing numerical errors dominate.

These changes have been tested using replay of https://1drv.ms/u/s!AunBK4rE1ZDCg_MeCKOBMYL9W4UIEQ?e=3b6oRq

Before:

Source log and replay show Z axis delta velocity variance collapses:

Screen Shot 2021-03-08 at 12 21 34 pm

Source log and replay show vertical position innovation diverging after variances collapse:
Screen Shot 2021-03-08 at 12 24 04 pm

Source log and replay show vertical velocity state variance collapses:
Screen Shot 2021-03-08 at 12 25 22 pm

After:

Replay Z axis delta velocity variance doesn't collapse:

https://1drv.ms/u/s!AunBK4rE1ZDCg_MeCKOBMYL9W4UIEQ?e=3b6oRq

Replay vertical position innovations don't diverge:

Screen Shot 2021-03-08 at 12 14 19 pm

Replay vertical velocity state variance doesn't collapse:

Screen Shot 2021-03-08 at 12 16 13 pm

It has also been tested on replay of logs submitted for https://github.com/ArduPilot/ardupilot/issues/16094. Refer to that issue for test results.

### CODE REVIEW

**Comment by tridge** [comment]
thanks Paul!

**Comment by Notorious811** [comment]
On it, have more than 20 quadcopters ready to bench test this PR, will soft reboot after every 45 minutes, try to create about 4-5 logs on each quadcopter, but i have backported this PR on ArduCopter v4.0.4, i have checked from my end and don't think backporting these changes on arducopter v4.0.4 will cause any conflicts

**Comment by Notorious811** [comment]
Completed the bench test as mentioned above, tested 22 quadcopters and kept them powered-on on the ground without any movements, for 14 quadcopters made multiple logs of 45 minutes each, and for rest 8 quadcopters made 2 logs of 2 hours each, after observing all the logs it seems like that this PR might has solved/handled a long standing issue related to EKF, which i also reported many times before as mentioned here https://discuss.ardupilot.org/t/continuous-uncommanded-climb-suspecting-ekf-issues/63137, https://github.com/ArduPilot/ardupilot/issues/15864, https://discuss.ardupilot.org/t/issues-related-to-imu-sensor/67740 and many other places like on discord

**Comment by priseborough** [comment]
> let me know if anyone willing to checkout logs for 22 quadcopters, i can share

@Notorious811  I'll look at those logs

**Comment by andyp1per** [comment]
@rmackay9 @priseborough this appears to fix the yaw drift issue for me!

**Comment by andyp1per** [comment]
Sorry, spoke too soon - tried a different copter and still seeing pronounced yaw drift.

**Comment by priseborough** [comment]
@andyp1per Can you please test this PR at commit https://github.com/ArduPilot/ardupilot/pull/16842/commits/dff864ae086e6b2b94b386a5bd92496a51d6da54 and let me know if that fixes your slow yaw spin issue when flying with just IMU and Baro.

**Comment by Notorious811** [comment]
@priseborough https://drive.google.com/file/d/1uxxOc45x4DB192RAUSqS8uGoI6qzy7z_/view?usp=sharing
Here are all the logs for 22 quadcopters, but these bench test logs are only for the changes made in this PR 3 days ago, i can see more modifications in this current PR, should i backport current changes too ? Older changes was simpler to backport to ArduCopter v4.0.4 which i am using currently and it didn't caused any conflicts as the pieces of codes which was changed in this PR earlier was last updated 4-5 years ago and those updates seems to solved my issue for now.

For the above logs, they all have same flight controller, custom build board ( almost a clone of omnibusf4pro ), but GPS are different, first 9 quadcopters having Ublox M8P GPS with EK3_ALT_SOURCE as GPS, and rest other quadcopters having Ublox M8N GPS with EK3_ALT_SOURCE as MS5611 barometer, few GPSes switch to GPS.status 2 or 1 which caused bad position estimations in very few of the logs, i guess thats due to cause i was bench testing all quadcopters closer to a building. For first 14 quadcopters, i armed the quadcopters without props at the end of all the logs.

**Comment by priseborough** [comment]
> should i backport current changes too ? 

@Notorious811  The additional changes were made in response to issues in other logs that were all related to issues with covariance updates when operating without aiding data for extended periods which was causing degraded yaw accuracy for operating without a magnetometer or learning bad IMU biases. I'm hoping those users can do the testing for those issues. 

Wait until I get a chance to check your logs before doing any further testing and until I get feedback on how the changes affect the issues reported in https://github.com/ArduPilot/ardupilot/issues/16094 as I may need to make further changes.

**Comment by andyp1per** [comment]
> 
> 
> @andyp1per Can you please test this PR at commit [dff864a](https://github.com/ArduPilot/ardupilot/commit/dff864ae086e6b2b94b386a5bd92496a51d6da54) and let me know if that fixes your slow yaw spin issue when flying with just IMU and Baro.

Will do

**Comment by andyp1per** [comment]
@priseborough yaw drift is gone 🎂 but I do appear to get a couple of alarming twitches after takeoff. I'll double check this is isn't the other changes I am flying with and get you a log if not.

**Comment by andyp1per** [comment]
Here's the log 
[log20.zip](https://github.com/ArduPilot/ardupilot/files/6136629/log20.zip)

**Comment by andyp1per** [comment]
@priseborough I think the twitch must have been the other changes - I haven't see it with just your changes. In general it works fine but a couple of times I got some drift (but not nearly as pronounced). When I switched replay on I could no longer reproduce. Here is a replay log in case it helps: 
[log4.zip](https://github.com/ArduPilot/ardupilot/files/6136918/log4.zip)

**Comment by priseborough** [comment]
@Notorious811 The vertical position innovation sequences in your logs (which show if the inertial nav is diverging away from the GPS or Baro) look OKwith the exception of some GPS height glitches which are unrelated to the original issue.

The key commits for your issue are:

https://github.com/ArduPilot/ardupilot/pull/16842/commits/96af55348b5e6d50b1b84a4343bc1e906ec42f7c
https://github.com/ArduPilot/ardupilot/pull/16842/commits/25b5f597eb0b90b1116850acf86d7ca10b7f7024

**Comment by Notorious811** [comment]
Thanks sir, I will definitely backport these commits and will do some more test flights in this week, provide some more results/logs by end of the week

**Comment by andyp1per** [comment]
Hi @priseborough is this ready to merge do you think?

**Comment by priseborough** [comment]
@andyp1per I've got nothing further I can do with it. It's over to the testers and reviewer.

**Review by andyp1per — APPROVED** [review]

**Comment by andyp1per** [comment]
Flew this successfully on a KakuteF7Mini copter

**Comment by andyp1per** [comment]
I flew with this today with no problems

File: libraries/AP_NavEKF3/AP_NavEKF3_AirDataFusion.cpp [hunk 1]
```diff
@@ -94,9 +94,14 @@ void NavEKF3_core::FuseAirspeed()
         }
 
         if (!inhibitDelVelBiasStates && !airDataFusionWindOnly) {
-            Kfusion[13] = SK_TAS[0]*(P[13][4]*SH_TAS[2] - P[13][22]*SH_TAS[2] + P[13][5]*SK_TAS[1] - P[13][23]*SK_TAS[1] + P[13][6]*vd*SH_TAS[0]);
-            Kfusion[14] = SK_TAS[0]*(P[14][4]*SH_TAS[2] - P[14][22]*SH_TAS[2] + P[14][5]*SK_TAS[1] - P[14][23]*SK_TAS[1] + P[14][6]*vd*SH_TAS[0]);
-            Kfusion[15] = SK_TAS[0]*(P[15][4]*SH_TAS[2] - P[15][22]*SH_TAS[2] + P[15][5]*SK_TAS[1] - P[15][23]*SK_TAS[1] + P[15][6]*vd*SH_TAS[0]);
+            for (uint8_t index = 0; index < 3; index++) {
+                uint8_t stateIndex = index + 13;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_AirDataFusion.cpp]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_AirDataFusion.cpp [hunk 2]
```diff
@@ -388,9 +393,14 @@ void NavEKF3_core::FuseSideslip()
         }
 
         if (!inhibitDelVelBiasStates && !airDataFusionWindOnly) {
-            Kfusion[13] = SK_BETA[0]*(P[13][0]*SK_BETA[5] + P[13][1]*SK_BETA[4] - P[13][4]*SK_BETA[1] + P[13][5]*SK_BETA[2] + P[13][2]*SK_BETA[6] + P[13][6]*SK_BETA[3] - P[13][3]*SK_BETA[7] + P[13][22]*SK_BETA[1] - P[13][23]*SK_BETA[2]);
-            Kfusion[14] = SK_BETA[0]*(P[14][0]*SK_BETA[5] + P[14][1]*SK_BETA[4] - P[14][4]*SK_BETA[1] + P[14][5]*SK_BETA[2] + P[14][2]*SK_BETA[6] + P[14][6]*SK_BETA[3] - P[14][3]*SK_BETA[7] + P[14][22]*SK_BETA[1] - P[14][23]*SK_BETA[2]);
-            Kfusion[15] = SK_BETA[0]*(P[15][0]*SK_BETA[5] + P[15][1]*SK_BETA[4] - P[15][4]*SK_BETA[1] + P[15][5]*SK_BETA[2] + P[15][2]*SK_BETA[6] + P[15][6]*SK_BETA[3] - P[15][3]*SK_BETA[7] + P[15][22]*SK_BETA[1] - P[15][23]*SK_BETA[2]);
+            for (uint8_t index = 0; index < 3; index++) {
+                uint8_t stateIndex = index + 13;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_AirDataFusion.cpp]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp [hunk 3]
```diff
@@ -602,9 +601,14 @@ void NavEKF3_core::FuseMagnetometer()
             }
 
             if (!inhibitDelVelBiasStates) {
-                Kfusion[13] = SK_MX[0]*(P[13][19] + P[13][1]*SH_MAG[0] - P[13][2]*SH_MAG[1] + P[13][3]*SH_MAG[2] + P[13][0]*SK_MX[2] - P[13][16]*SK_MX[1] + P[13][17]*SK_MX[4] - P[13][18]*SK_MX[3]);
-                Kfusion[14] = SK_MX[0]*(P[14][19] + P[14][1]*SH_MAG[0] - P[14][2]*SH_MAG[1] + P[14][3]*SH_MAG[2] + P[14][0]*SK_MX[2] - P[14][16]*SK_MX[1] + P[14][17]*SK_MX[4] - P[14][18]*SK_MX[3]);
-                Kfusion[15] = SK_MX[0]*(P[15][19] + P[15][1]*SH_MAG[0] - P[15][2]*SH_MAG[1] + P[15][3]*SH_MAG[2] + P[15][0]*SK_MX[2] - P[15][16]*SK_MX[1] + P[15][17]*SK_MX[4] - P[15][18]*SK_MX[3]);
+                for (uint8_t index = 0; index < 3; index++) {
+                    uint8_t stateIndex = index + 13;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_RngBcnFusion.cpp [hunk 4]
```diff
@@ -149,9 +149,14 @@ void NavEKF3_core::FuseRngBcn()
         }
 
         if (!inhibitDelVelBiasStates) {
-            Kfusion[13] = -t26*(P[13][7]*t4*t9+P[13][8]*t3*t9+P[13][9]*t2*t9);
-            Kfusion[14] = -t26*(P[14][7]*t4*t9+P[14][8]*t3*t9+P[14][9]*t2*t9);
-            Kfusion[15] = -t26*(P[15][7]*t4*t9+P[15][8]*t3*t9+P[15][9]*t2*t9);
+            for (uint8_t index = 0; index < 3; index++) {
+                uint8_t stateIndex = index + 13;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_RngBcnFusion.cpp]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_core.cpp [hunk 5]
```diff
@@ -1136,6 +1133,24 @@ void NavEKF3_core::CovariancePrediction(Vector3f *rotVarVecPtr)
     float _accNoise = constrain_float(frontend->_accNoise, 0.0f, 10.0f);
     dvxVar = dvyVar = dvzVar = sq(dt*_accNoise);
 
+    if (!inhibitDelVelBiasStates) {
+        for (unsigned stateIndex = 13; stateIndex <= 15; stateIndex++) {
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.cpp]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_core.cpp [hunk 6]
```diff
@@ -1664,6 +1679,18 @@ void NavEKF3_core::CovariancePrediction(Vector3f *rotVarVecPtr)
         }
     }
 
+    // inactive delta velocity bias states have all covariances zeroed to prevent
+    // interacton with other states
+    if (!inhibitDelVelBiasStates) {
+        for (uint8_t index=0; index<3; index++) {
+            uint8_t stateIndex = index + 13;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.cpp]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_core.h [hunk 7]
```diff
@@ -1325,6 +1324,11 @@ class NavEKF3_core : public NavEKF_core_common
     bool onGroundNotMoving;             // true when on the ground and not moving
     uint32_t lastMoveCheckLogTime_ms;   // last time the movement check data was logged (msec)
 
+	// variables used to inhibit accel bias learning
+    bool inhibitDelVelBiasStates;       // true when all IMU delta velocity bias states are de-activated
+	bool dvelBiasAxisInhibit[3] {};		// true when IMU delta velocity bias states for a specific axis is de-activated
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_core.h]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_OptFlowFusion.cpp [hunk 8]
```diff
@@ -466,9 +466,14 @@ void NavEKF3_core::FuseOptFlow()
             }
 
             if (!inhibitDelVelBiasStates) {
-                Kfusion[13] = t78*(P[13][0]*t2*t5-P[13][4]*t2*t7+P[13][1]*t2*t15+P[13][6]*t2*t10+P[13][2]*t2*t19-P[13][3]*t2*t22+P[13][5]*t2*t27);
-                Kfusion[14] = t78*(P[14][0]*t2*t5-P[14][4]*t2*t7+P[14][1]*t2*t15+P[14][6]*t2*t10+P[14][2]*t2*t19-P[14][3]*t2*t22+P[14][5]*t2*t27);
-                Kfusion[15] = t78*(P[15][0]*t2*t5-P[15][4]*t2*t7+P[15][1]*t2*t15+P[15][6]*t2*t10+P[15][2]*t2*t19-P[15][3]*t2*t22+P[15][5]*t2*t27);
+                for (uint8_t index = 0; index < 3; index++) {
+                    uint8_t stateIndex = index + 13;
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_OptFlowFusion.cpp]: fixed

File: libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp [hunk 9]
```diff
@@ -244,10 +244,9 @@ void NavEKF3_core::SelectMagFusion()
                 if (onGroundNotMoving) {
                     // fuse last known good yaw angle before we stopped moving to allow yaw bias learning when on ground before flight
                     fuseEulerYaw(yawFusionMethod::STATIC);
-                } else if (onGround || (sq(P[0][0])+sq(P[1][1])+sq(P[2][2])+sq(P[3][3]) > 0.01f)) {
-                    // prevent uncontrolled yaw variance growth by fusing a zero innovation
-                    // when not on ground allow more variance growth so yaw can be corrected
-                    // by manoeuvring
+                } else if (onGround || PV_AidingMode == AID_NONE || (P[0][0]+P[1][1]+P[2][2]+P[3][3] > 0.01f)) {
```
[the remark this answers is not in the document]
Comment (priseborough) [reply libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp:247]: Yes it was intentional.  The original intention was to allow the angular uncertainty to grow to about a 0.1 rad uncertainty which is a variance of 0.01 rad^2. The P covariance matrix diagonals are already variances and do not have to be squared.

**Review by tridge — APPROVED** [review]

**Comment by LeiOuYang** [comment]
@priseborough @tridge HI! I used the new version of the firmware(Copter4.2.0-dev) to test on the ground and the yaw was always drift, but it was difficult to reproduce it later. I attached the log:
https://1drv.ms/u/s!AtY6Wtn1T07BcfU5bEg7ZWY1JZY?e=7djaWp
I hope you can see it sometime

**Pull request merged**